raft.go

```diff
@@ -914,19 +914,44 @@ func (r *raft) hup(t CampaignType) {
 		r.logger.Warningf("%x is unpromotable and can not campaign", r.id)
 		return
 	}
-	ents, err := r.raftLog.slice(r.raftLog.applied+1, r.raftLog.committed+1, noLimit)
-	if err != nil {
-		r.logger.Panicf("unexpected error getting unapplied entries (%v)", err)
-	}
-	if n := numOfPendingConf(ents); n != 0 && r.raftLog.committed > r.raftLog.applied {
-		r.logger.Warningf("%x cannot campaign at term %d since there are still %d pending configuration changes to apply", r.id, r.Term, n)
+	if r.hasUnappliedConfChanges() {
+		r.logger.Warningf("%x cannot campaign at term %d since there are still pending configuration changes to apply", r.id, r.Term)
 		return
 	}
 
 	r.logger.Infof("%x is starting a new election at term %d", r.id, r.Term)
 	r.campaign(t)
 }
 
+func (r *raft) hasUnappliedConfChanges() bool {
+	if r.raftLog.applied >= r.raftLog.committed { // in fact applied == committed
+		return false
+	}
+	// Scan all unapplied committed entries to find a config change. Paginate the
+	// scan, to avoid a potentially unlimited memory spike.
+	//
+	// TODO(pavelkalinnikov): this scan does not need to happen each time. It's
+	// possible to scan once to bootstrap, and then maintain the "has unapplied
+	// config changes" predicate throughout the lifetime of this instance, as all
+	// new log entries pass through "for free".
+	const maxSize = entryEncodingSize(16 << 20) // 16 MiB
+	for idx, end := r.raftLog.applied+1, r.raftLog.committed+1; idx < end; {
+		ents, err := r.raftLog.slice(idx, end, maxSize)
+		if err != nil {
+			r.logger.Panicf("unexpected error getting unapplied entries from %d: %v", idx, err)
+		} else if len(ents) == 0 {
+			r.logger.Panicf("could not read unapplied entries in [%d, %d)", idx, end)
+		}
+		for i := range ents {
+			if ents[i].Type == pb.EntryConfChange || ents[i].Type == pb.EntryConfChangeV2 {
+				return true
+			}
+		}
+		idx += uint64(len(ents))
+	}
+	return false
+}
```

Comment:
The pagination doesn't make sense. Actually the memory has already been allocated in `(*raftLog) slice`, and it just returns a slice on top a underlying array. 

I would suggest to simplify the method something like below,
```
func (r *raft) hasUnappliedConfChanges() bool {
	if r.raftLog.applied >= r.raftLog.committed { // in fact applied == committed
		return false
	}

	ents, err := r.raftLog.slice(r.raftLog.applied+1, r.raftLog.committed+1, noLimit)
	if err != nil {
		r.logger.Panicf("unexpected error getting unapplied entries from (%v)", err)
	}
	for i := range ents {
		if ents[i].Type == pb.EntryConfChange || ents[i].Type == pb.EntryConfChangeV2 {
			return true
		}
	}

	return false
}

```

Reply:
> `raftLog.slice` fetches a prefix of entries [from storage](https://github.com/etcd-io/raft/blob/2e0653d20750c7cbbf8473fd1582c63a867996c1/log.go#L470-L471), and a suffix of entries from memory (in the `unstable` type). The storage fetching part can read arbitrary amount of data into memory if the `committed..applied` range is large, and/or if there are large (in bytes) entries in this range. Pagination protects from this scenario.

Right, the key point is that `(*raftLog) slice` might reallocate memory at [log.go#L490](https://github.com/etcd-io/raft/blob/7699239b357a3acf26860eb6683fdcc788b07349/log.go#L490). I am thinking we need to skip [log.go#L487-L497](https://github.com/etcd-io/raft/blob/7699239b357a3acf26860eb6683fdcc788b07349/log.go#L487-L497) if [ents](https://github.com/etcd-io/raft/blob/7699239b357a3acf26860eb6683fdcc788b07349/log.go#L485) already hits the `maxSize`. If `len(ents) < cap(ents)`, then it means it already hits the `maxSize`.  

We might also need to change

https://github.com/etcd-io/raft/blob/7699239b357a3acf26860eb6683fdcc788b07349/storage.go#L133

to 
```
ents := ms.ents[lo-offset : hi-offset: hi-offset]
```

Reply:
Can we merge https://github.com/etcd-io/raft/pull/34 firstly? Afterwards, rebase this PR, then we can take a look again.

Reply:
@tbg I like the idea of `paginate` method. It also allows avoiding the `append` of the `unstable` in `raftLog.slice` - we can just return the `unstable` in its own page.

Reply:
Here is the same experiment as above but with this PR. You can see that it solves the problem, memory usage doesn't spike at all.

![image](https://github.com/etcd-io/raft/assets/5076964/31d66e31-88a5-4ebe-8a8a-6f7e5a836f55)

Reply:
Thanks for the explanation.  My original comment isn't valid anymore after merging a couple of other PRs.

Reply:
Added a `scan` method. Put the memory usage graphs to the PR description.